client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java

```diff
@@ -187,7 +187,9 @@ private void failConnection(Throwable throwable) {
                     log.warn("Connection for segment {} failed due to: {}", segmentName, throwable.getMessage());
                 }
             }
-            if (failSetupConnection) {
+            if (throwable instanceof SegmentSealedException || throwable instanceof NoSuchSegmentException) {
```

Comment:
why do we have to special case NoSuchSegmentException. 
What is the exit condition for other exceptions.. does the writer keep retrying to connect?

Reply:
These are being special cased because they are not recoverable. If a segment does not exist or is sealed we assume there is no point in reconnecting with backoff hoping that will change. So these are being set so that future callers to write will fail with the exception, and not resetting so that it does not attempt to reconnect.